frame/vesting/src/lib.rs

```diff
@@ -302,72 +409,273 @@ pub mod pallet {
+		/// NOTE: If both schedules have ended by the current block, no new schedule will be created.
+		///
+		/// Merged schedule attributes:
+		/// starting_block: `MAX(schedule1.starting_block, scheduled2.starting_block, current_block)`.
+		/// ending_block: `MAX(schedule1.ending_block, schedule2.ending_block)`.
+		/// locked: `schedule1.locked_at(current_block) + schedule2.locked_at(current_block)`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `schedule1_index`: index of the first schedule to merge.
+		/// - `schedule2_index`: index of the second schedule to merge.
+		///
+		/// # <weight>
+		/// - `O(1)`.
+		/// - DbWeight: TODO Reads, TODO Writes
+		///     - Reads: TODO
+		///     - Writes: TODO
+		/// # </weight>
+		#[pallet::weight(
+			T::WeightInfo::not_unlocking_merge_schedules(MaxLocksOf::<T>::get())
+			.max(T::WeightInfo::unlocking_merge_schedules(MaxLocksOf::<T>::get()))
+		)]
+		pub fn merge_schedules(
+			origin: OriginFor<T>,
+			schedule1_index: u32,
+			schedule2_index: u32,
+		) -> DispatchResult {
+			if schedule1_index == schedule2_index {
+				return Ok(());
+			};
+			let who = ensure_signed(origin)?;
+			let schedule1_index = schedule1_index as usize;
+			let schedule2_index = schedule2_index as usize;
+			let vesting = Self::vesting(&who).ok_or(Error::<T>::NotVesting)?;
+			let len = vesting.len();
+			ensure!(
+				schedule1_index < len && schedule2_index < len,
+				Error::<T>::ScheduleIndexOutOfBounds
+			);
+
+			// The schedule index is based off of the schedule ordering prior to filtering out any
+			// schedules that may be ending at this block.
+			let schedule1 = vesting[schedule1_index];
+			let schedule2 = vesting[schedule2_index];
+			let filter = Filter::Two((schedule1_index, schedule2_index));
+
+			// The length of vesting decreases by 2 here since we filter out 2 schedules. Thus we know
+			// below that we can safely insert the new merged schedule.
+			let maybe_vesting = Self::update_lock_and_schedules(who.clone(), vesting, filter);
+
+			// We can't fail from here on because we have potentially removed two schedules.
 
-			Self::add_vesting_schedule(&target, schedule.locked, schedule.per_block, schedule.starting_block)
-				.expect("user does not have an existing vesting schedule; q.e.d.");
+			let now = <frame_system::Pallet<T>>::block_number();
+			if let Some(s) = Self::merge_vesting_info(now, schedule1, schedule2) {
+				let mut vesting = maybe_vesting.unwrap_or_default();
+				// It shouldn't be possible for this to fail because we removed 2 schedules above.
+				ensure!(vesting.try_push(s).is_ok(), Error::<T>::AtMaxVestingSchedules);
+				Self::deposit_event(Event::<T>::VestingMergeSuccess(
+					s.locked(),
+					s.per_block(),
+					s.starting_block(),
+				));
+				Vesting::<T>::insert(&who, vesting);
+			} else if maybe_vesting.is_some() {
+				Vesting::<T>::insert(&who, maybe_vesting.unwrap());
```

Comment:
```suggestion
				Vesting::<T>::insert(&who, maybe_vesting.expect("checked above; qed));
```

Always `expect`, never `unwrap`.

Reply:
this is an anti-pattern anyway.

Should be:

```
if let Some(vesting) = maybe_vesting {
    Vesting::<T>::insert(&who, vesting);
}

Reply:
afe25d2